### PR TITLE
AWS "EC2 instance" termination_protection and source_dest_check changeability at run-time + block_device_mapping output

### PR DESCRIPTION
**Pull request by amandolo**
- Added "EC2 instance" termination_protection and source_dest_check changeability at run-time
- More detailed "block_device_mapping" informations

### CODE REVIEW

**Comment by amandolo** [comment]
Imported from previous pull request (https://github.com/ansible/ansible/pull/8144)

**Comment by bobobox** [comment]
+1 on this. Currently we are having to issue an extra aws cli command to get this done, which is not so nice.

**Comment by bobobox** [comment]
Now using the PR's version of the module via playbooks/library/ and it's working well.

**Comment by juliedavila** [comment]
@amandolo can you rebase plz?

**Comment by amandolo** [comment]
Done !! ;)

**Comment by juliedavila** [comment]
@amandolo last thing we need here is to avoid having merge commits in your pull request.

Currently your merge commits are:
1. bbf719d9155548a50fa6919a707c2cfe7ea8458e
2. c5cafa250af0f6c9b6f208d0d5ab5e3eeeac4c40
3. 0ffe37cebe22f454ae288035846461e228318b86

For future ref plz use “git pull –rebase” vs “git pull” and “git rebase” vs “git merge” which should help you avoid the above.

File: cloud/amazon/ec2.py [hunk 1]
```diff
@@ -1120,21 +1145,38 @@ def startstop_instances(module, ec2, instance_ids, state):
...
+                    changed = True
+            except Exception as error:
+                pass
+
+            # Check "termination_protection" attribute
+            try:
+                if inst.get_attribute('disableApiTermination')['disableApiTermination'] != termination_protection:
+                    inst.modify_attribute('disableApiTermination', termination_protection)
+                    changed = True
+            except Exception as error:
```
Comment (juliedavila) [inline cloud/amazon/ec2.py]: Why are you passing the exception here?

File: cloud/amazon/ec2.py [hunk 2]
```diff
@@ -1120,21 +1145,38 @@ def startstop_instances(module, ec2, instance_ids, state):
...
-               except EC2ResponseError, e:
-                   module.fail_json(msg='Unable to change state for instance {0}, error: {1}'.format(inst.id, e))
-               changed = True
+
+            # Check "source_dest_check" attribute
+            try:
+                if inst.get_attribute('sourceDestCheck')['sourceDestCheck'] != source_dest_check:
+                    inst.modify_attribute('sourceDestCheck', source_dest_check)
+                    changed = True
+            except Exception as error:
```
Comment (juliedavila) [inline cloud/amazon/ec2.py]: same as the comment on 1166

**Comment by nritholtz** [comment]
+1

**Comment by raags** [comment]
@amandolo can you also add a `disable_protection_for_termination` so that the termination protection can be disabled when instance is being removed.

**Comment by amandolo** [comment]
I have rebased the outdated branch.

**Comment by amandolo** [comment]
@raags you can do this in 2 steps; with my edits you can change termination_protection attribute on the fly. I think this isn't a good idea for a "secure" reason.

**Comment by blandonnimrat** [comment]
+1 this would be most excellent.

**Comment by cove** [comment]
+1

**Comment by bcoca** [comment]
this is still a problem to merge

**Comment by raags** [comment]
@amandolo in the task is to remove an instance, the option to disable termination protection is explicit. The default is false, so when you enable it, you are specific regarding your requirement. I think one task ought to do this, instead of two.

**Comment by gregdek** [comment]
Yup, still needs to be rebased.

**Comment by amandolo** [comment]
Branch rebased.

File: cloud/amazon/ec2.py [hunk 3]
```diff
@@ -613,6 +620,19 @@ def get_instance_info(inst):
         instance_info['ebs_optimized'] = False
 
     try:
+        bdm_dict = {}
```
Comment (jimi-c) [inline cloud/amazon/ec2.py:623]: What do these lines have to do with the described new features? This seems to be bleed-over from another feature you're working on?

Comment (amandolo) [inline cloud/amazon/ec2.py:623]: Indeed you are right, i should have to open another merge request instead to insert this here.
But this is ONLY an extra output for some info of any attached disk.

Do I have to separate the requests ?

Comment (jimi-c) [inline cloud/amazon/ec2.py:623]: Go ahead and make sure the new feature is fully represented in this PR and update the commit message, and we can merge them all in one shot.

**Comment by amandolo** [comment]
Fix commits to separate different features.

**Comment by bennojoy** [comment]
@amandolo  one more change can you make version_added: 2.0 for the new param thanks..

**Comment by amandolo** [comment]
Done!

**Comment by bennojoy** [comment]
Merged Thanks..

**Pull request merged**